scanpy/external/tl/_sam.py

```diff
@@ -0,0 +1,216 @@
+        recommend using 'StandardScaler' for large datasets with many
+        expected cell types and 'Normalizer' otherwise.
+
+    num_norm_avg - int, optional, default 50
+        The top 'num_norm_avg' dispersions are averaged to determine the
+        normalization factor when calculating the weights. This prevents
+        genes with large spatial dispersions from skewing the distribution
+        of weights.
+
+    weight_pcs - bool, optional, default True
+        If True, scale the principal components by their eigenvalues. In
+        datasets with many expected cell types, setting this to False might
+        improve the resolution as these cell types might be encoded by low-
+        variance principal components.
+
+    npcs - int, optional, default None,
+        Determines the number of top principal components selected at each
+        iteration of the SAM algorithm. If None, this number is chosen
+        automatically based on the size of the dataset. If weight_pcs is
+        set to True, this parameter primarily affects the runtime of the SAM
+        algorithm (more PCs = longer runtime).
+
+    n_genes - int, optional, default None:
+        Determines the number of top SAM-weighted genes to use at each iteration
+        of the SAM algorithm. If None, this number is chosen automatically
+        based on the size of the dataset. This parameter primarily affects
+        the runtime of the SAM algorithm (more genes = longer runtime).
+    
+    inplace - bool, optional, default True:
+        Set fields in `adata` if True. Otherwise, returns a copy.
+    
+    verbose - bool, optional, default True:
+        If True, displays SAM log statements.
+
+    Returns
+    -------
+    sam - SAM
+        The SAM object
+
+    adata - AnnData
+        `.var['weights']`
+            SAM weights for each gene.
+        `.var['spatial_dispersions']`
+            Spatial dispersions for each gene (these are used to compute the
+            SAM weights)
+        `.var['mask_genes']`
+            If preprocessed with SAM, this boolean vector indicates which genes
+            were filtered out (=False).
+        `.uns['preprocess_args']`
+            Dictionary of parameters used for preprocessing.
+        `.uns['run_args']`
+            Dictionary of parameters used for running SAM.
+        `.uns['pca_obj']`
+            The sklearn.decomposition.PCA object.
+        `.uns['X_processed']`
+            The standardized and SAM-weighted data fed into PCA.
+        `.uns['neighbors']`
+            A dictionary with key 'connectivities' containing the kNN adjacency
+            matrix output by SAM. If built-in scanpy dimensionality reduction
+            methods are to be used using the SAM-output AnnData, users
+            should recompute the neighbors using `.obs['X_pca']` with
+            `scanpy.pp.neighbors`.
+        `.uns['ranked_genes']`
+            Gene IDs ranked in descending order by their SAM weights.
```

Comment:
Shouldn’t this be `.var['sam_rank']` or so?

Reply:
It's the genes themselves sorted in decreasing order of the SAM weights, but it's not really useful since you can just sort `.var['weights']` to get the same information. I removed it!

This is kind of hacky, but to avoid having to refactor all my pipelines/scripts by changing what gets output into `uns` by `SAM`, I incorporated your suggestions by manually changing `uns` after `s.run(...)`:

```
    s.adata.uns['sam'] = {'run_args' : s.adata.uns['run_args'],
     'preprocess_args' : s.adata.uns['preprocess_args']}

    del s.adata.uns['nnm']
    del s.adata.uns['preprocess_args']
    del s.adata.uns['run_args']
    del s.adata.uns['ranked_genes']
```
Hopefully, this is an acceptable compromise!

Reply:
This is probably a more elegant way of writing the above:
```
    s.adata.uns['sam'] = {}
    for attr in ['nnm','preprocess_args','run_args','ranked_genes']:
        s.adata.uns['sam'][attr] = s.adata.uns.pop(attr,None)
```